### ISSUE TITLE
`AzureSearch` vectorstore should be converted to `AzureSearchVectorStoreRetriever` when calling `as_retriever`

### ISSUE THREAD

**Issue by mspronesti** [langchain-ai/langchain#20600]
### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the LangChain documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangChain rather than my code.
- [X] The bug is not resolved by updating to the latest stable version of LangChain (or the specific integration package).

### Example Code

```py
openai_api_version = "2024-02-01"

embeddings = AzureOpenAIEmbeddings(
    deployment=os.getenv('EMBEDDING_DEPLOYMENT_NAME'),
    openai_api_version=openai_api_version,
)

index_name: str = "my-index-name"

fields = [...]

vector_store: AzureSearch = AzureSearch(
    azure_search_endpoint=os.getenv('VECTOR_STORE_ADDRESS'),
    azure_search_key=os.getenv('VECTOR_STORE_PASSWORD'),
    index_name=index_name,
    embedding_function=embedding_function,
    fields=fields,
    # needed for semantic ranking
    semantic_configuration_name = 'my-config',
)

retriever = vector_store.as_retriever(search_kwargs={"k": 5}, search_type="semantic_hybrid")
```

### Description

I'm trying to use "semantic_hybrid" as `search_type` but it is not supported, as `as_retriever` doesn't return an instance of the class `AzureSearchVectorStoreRetriever` but a `VectorStoreRetriever`.

### System Info

-

**Comment by chrislrobert** [comment]
If anybody comes across this fix and has a similar issue, this is my updated override code to essentially undo the fix:

```
class ExtendedAzureSearch(AzureSearch):
    """Extended AzureSearch class with support for similarity_score_threshold searches."""

    def as_retriever(self, **kwargs: Any) -> VectorStoreRetriever:
        """Return VectorStoreRetriever initialized from this VectorStore. Hard-wired to return base class version."""

        # (the latest version in AzureSearch triggers an invalid search_type error)
        return VectorStore.as_retriever(self, **kwargs)

    def _select_relevance_score_fn(self) -> Callable[[float], float]:
        return self._azure_relevance_score_fn

    @staticmethod
    def _azure_relevance_score_fn(score: float) -> float:
        """Returns Azure's similarity score as-is."""

        # see https://learn.microsoft.com/en-us/azure/search/vector-search-ranking#scores-in-a-vector-search-results
        return score

    def similarity_search_with_score(
        self, *args: Any, **kwargs: Any
    ) -> List[Tuple[Document, float]]:
        """Run similarity search with distance."""

        # bridge into the AzureSearch class to run similarity search with distance
        return self.vector_search_with_score(*args, **kwargs)
```

**Comment by ccurme** [comment]
Hi @chrislrobert,

We'd like to fix this. I opened a draft PR here, which is entirely untested: https://github.com/langchain-ai/langchain/pull/20837

Would something like this fix your use-case?

Happy to update (or review a contribution from you and close the PR) if it's off the mark. Thanks.

**Comment by mspronesti** [comment]
@ccurme If I got it right, @chrislrobert had to tweak the internals of `AzureSearch` to make it work for his use case even before fixing `as_retriever`, hence I don't see this as a breaking change. To me, it would make sense to do the other way round, i.e. given an `AzureSearch` object

```py
VectorStoreRetriever(vectorstore=vectorstore)
```
What do you guys think? :)

**Comment by chrislrobert** [comment]
Thanks, @ccurme, for looking into this! I guess a few comments:

1. Your PR looks like it basically reverses this change, which does fix my use case
2. I agree with @mspronesti that my specific use case was potentially less important given that I already had to override the AzureSearch class to get my case to work
3. However, I think the broader issue is that the base class is currently providing a lot of useful functionality — including most of the necessary support for `similarity_score_threshold` searches, plus, it sounds like, `async` support — and the change limits the extent to which people can benefit from that base-class support
4. If it's true that this effectively broke `async` support as well (provided by the base class), then that's much more serious
5. The cleanest fix would probably be to fill out explicit `similarity_score_threshold` and `async` support in the AzureSearch classes so that changes like this one here pose no issues and the relied-upon functionality can be sort of brought out of the shadows

Either way, I appreciate the quick attention!

**Comment by mspronesti** [comment]
@chrislrobert If I got it right, you are looking for a way to retrieve the most relevant docs using "similarity" as a metric, **plus the score associated to each doc**. Is this right? 

E.g.
```py
vectorstore = AzureSearch(...) 
retriever = vectorstore.as_retriever(search_kwargs={"k": 5, "search_type": "similarity_score_threshold"})

# use the retriever here 
# ...
```

**Comment by chrislrobert** [comment]
@mspronesti Basically, yes: the trouble with a fixed `k` is that even just "Hi!" can get you `k` totally-irrelevant results... so it's helpful to set a minimum threshold score for inclusion (as documented in https://python.langchain.com/docs/modules/data_connection/retrievers/vectorstore/#similarity-score-threshold-retrieval). And you're right, you need to also pass the threshold as a parameter, as in:

```
        conversation_chain = ConversationalRetrievalChain(
            retriever=vectorstore.as_retriever(search_type="similarity_score_threshold",
                                               search_kwargs={"score_threshold": rag_score_threshold, "k": rag_top_k}),
        ...
        )
```

**Comment by mspronesti** [comment]
@chrislrobert I've just opened #20852. Could you please give it a look and let me know if that helps? :)

**Comment by DhavalThkkar** [comment]
@chrislrobert I guess some changes have cause my code to break with respect to ConfigurableFields. Here is the issue [#21533 ](https://github.com/langchain-ai/langchain/discussions/21533)

Can someone maybe help with this? It used to work in an earlier version of langchain-community

**Comment by mspronesti** [comment]
@chrislrobert @DhavalThkkar This issue, and the subsequent PR I submitted, aimed at returning an instance of `AzureSearchVectorStoreRetriever` with `as_retriever` instead of a "regular" `VectorStoreRetriever`. This was needed because Azure AI Search could handle some additional search types thus somebody thought of designing a custom retriever to handle those. 

Nevertheless, I didn't look inside such implementation -- I assumed it was proper. It turned out it had some bugs, e.g. it didn't support `similarity_score_threshold` (hopefully resolved in #20601) and it didn't make use of `search_kwargs`. For the latter, there's a PR open (#21572). Feel free to give it a look.
